### DISCUSSION TITLE
How to provide independent non-interfering subject filtering rules for multiple domain names

### DISCUSSION THREAD

**Discussion by eosps** [rspamd/rspamd#4582] [General]
I have multiple email domains. I want to configure different subject checking strategies for each domain name, and they do not affect each other. I configured multimap but it doesn't seem to work. Please help me to see where the configuration is wrong.

----->multimap.conf

MARK_SUBJECT_DOMAIN_CHEAK {
     type = "header";
     header = "Subject";
     filter = regexp:/re/;
     selector = rcpts('smtp'):domain.lower;
     map = "${LOCAL_CONFDIR}/local.d/maps.d/subject_${selector}.map";
     prefilter = true;
     description = "cheak subject";
     regexp = true;
     score = 8.0;
}

**Comment by vstakhov** [comment]
This is not an issue but a support question

**Comment by eosps** [comment]
My configuration doesn't work. I want to know the correct way. Or is there any other method, similar to the dkim configuration method, to select different maps according to different domain names?

**Comment by eosps** [comment]
It would be nice to have a selector like dkim. There is a path_map option to choose the map files prepared in advance to filter separately.
I can't use redis either. The configuration is as follows:

MARK_SUBJECT_DOMAIN_CHEAK {
     type = "header";
     header = "Subject";
     filter = regexp:/re/;
     selector = rcpts('smtp'):domain.lower;
     #map = "${LOCAL_CONFDIR}/local.d/maps.d/subject_${selector}.map";
     map = "redis+selector://subject_${selector}";
     prefilter = true;
     description = "cheak subject";
     regexp = true;
     #action = "reject";
     #action = "add header";
     score = 8.0;
}

**Comment by eosps** [comment]
### I wrote a Lua script but it doesn't seem to work. Please help me. If the map corresponding to the recipient's domain name exists, compare the subject with the regular expression in the map. If the comparison is successful, it will score 8 points.
---------------------------------------------------------------------------------------------------------

      local rspamd_logger = require "rspamd_logger"
      local lua_maps = require "lua_maps"
      local function check_subject_regex(task)
      
          local recipients = task:get_recipients()
          local domain_scores = {}  
          for _, recipient in ipairs(recipients) do
              local domain = recipient['domain']
              domain_scores[domain] = 0.0
          end
          for domain, _ in pairs(domain_scores) do
              rspamd_logger.infox(task, "------>: %s", domain)
              local filter_table = lua_maps.map_add_from_ucl(
                  'file:///usr/share/rspamd/etc/local.d/maps.d/subject_' .. domain .. '.map',
                  'regexp',
                  'Filter table for ' .. domain
              )
              --rspamd_logger.infox(task, "------>: %s", filter_table)
              local filter_expression = filter_table:get_key(domain)
              --rspamd_logger.infox(task, "------>: %s", type(filter_expression))
              if filter_expression then
                  local filter_pattern = rspamd_regexp.create(filter_expression)
                  local subject = task:get_header('Subject')
                  --rspamd_logger.infox(task, "------>: %s", subject)
                  if filter_pattern:match(subject) then
                      return true
                  end
              end
          end
          return false
      end
      
      rspamd_config:register_symbol({
          name = ' MARK_SUBJECT_DOMAIN_CHEAK',
          type = 'pre_filter',
          scores = 8.0,
          callback = check_subject_regex,
          priority = 3
      })


### The program seems to have a problem at 
“ local filter_expression = filter_table:get_key(domain) ”

**Reply by vstakhov** [reply]
Your script has multiple issues and will not work (it will also use all memory and crash Rspamd eventually). Please check the code of the existing rules/plugins and read the documentation about how to write rules.

**Comment by eosps** [comment]
I have no experience in writing Lua, and I am still figuring it out. thanks for your reminder
